content/develop/concepts/architecture/fragments

```diff
@@ -0,0 +1,143 @@
+---
+title: Working withwq fragments and partial reruns
+slug: /develop/concepts/architecture/fragments
+---
+
+# Working with fragments and partial reruns
+
+Reruns are a central part of every Streamlit app. When users interact with widgets, your script reruns from top to bottom and your app's frontend is updated. Streamlit provides several features to help you develop your app within this execution model. Fragments were introduced in Streamlit version 1.33.0 to allow rerunning a portion or your code instead of your full script. As your app grows larger and more complex, partial reruns help your app efficient and performant.
+
+Before you read about fragments, we recommend having a basic understanding of Session State and forms.
+
+## Use cases for fragments
+
+Fragments are versatile and applicable to a wide variety of circumstances. Here are just a few, common scenarios where fragments are useful:
+
+- Your app has multiple visualizations and each one takes time to load, but you have a filter input that only updates one of them.
+- You have a dynamic form that doesn't need to update the rest of your app (until the form is complete).
+- You want to automatically update a single component or group of components to stream data.
+
+## Defining and calling a fragment
+
+Streamlit provides a decorator to turn any function into a fragment function. When you call a fragment function that contains a widget, a user triggers a _partial rerun_ instead of a full rerun when they interact with that fragment's widget. During a partial rerun, your fragment function is re-executed. Anything within the main body of your fragment is updated on the frontend while the rest of your app remains the same. We will discuss what happens when you write to other containers within a fragment later on.
+
+Here is a basic example of defining and calling a fragment function:
+
+```ptyhon
+import streamlit as st
+
+@st.experimental_fragment
+def fragment_function():
+    if st.button("Hi!"):
+        st.write("Hi back!")
+
+fragment_function()
+```
+
+If you want the main body of your fragment to appear in the sidebar or another container, call your fragment function inside a context manager.
+
+```python
+with st.sidebar:
+    fragment_function()
+```
+
+### Partial rerun execution flow
+
+Consider the following code with the explanation and diagram that follow.
+
+```python
+import streamlit as st
+
+st.title("My Awesome App")
+
+@st.experimental_fragment()
+def fragment_one():
+    cols = st.columns(2)
+    cols[0].toggle("Toggle")
+    cols[1].text_area("Enter text")
+
+@st.experimental_fragment()
+def fragment_two():
+    cols = st.columns(2)
+    cols[0].checkbox("Filter")
+    cols[1].file_uploader("Upload image")
```

Comment:
Nit: Does it make sense to define this function below, before it's executed, instead of defining at the top? IMO it makes the app code a bit harder to read when it's separated. Also acknowledging that folks will copy whatever we do here :)

Reply:
I'm happy to do things either way for this example. In principle, I tend to prefer using description function names and defining them separately from the main body of the code so they can be understood and containerized in the mind as you're reading through a larger process. That's not particularly relevant in this toy example, but that's why I defaulted to that arrangement.

Reply:
OK, can I request either:
- Change the function names to something descriptive, even like `toggle_and_text()` and `filter_and_file()`, OR
- Rearrange so the functions are defined in line and the st commands are read in the same order they are rendered